resources/tools/check-markdown/src/check-link.ts

```diff
@@ -0,0 +1,59 @@
+import * as fs from 'node:fs';
+import { posix as path } from 'node:path';
+
+import { Tokens } from 'marked';
+
+import { LinkRef, LinkRefMessage } from './types.js';
+
+export function checkLinks(root: string, links: LinkRef[]) {
+  let result = true;
+  for(const file of links) {
+    for(const link of file.links) {
+      // verify that the file exists in filesystem at expected location
+      result = checkLink(root, file.file, <Tokens.Link>link, file.messages) && result;
+    }
+  }
+  return result;
+}
+
+function checkLink(root: string, file: string, token: Tokens.Link | Tokens.Image, messages: LinkRefMessage[]) {
+  const parsed = token.href.split('#');
+  const href = parsed[0];
+  // const anchor = parsed.length > 1 ? parsed[1] : '';
+
+  if(href.startsWith('https:') || href.startsWith('http:')) {
+    messages.push({token, type:'info', message: 'External link'});
+    return true;
+
+  }
+  if(href.startsWith('/')) {
+    messages.push({token, type:'info', message: 'Absolute path'});
+    return true;
+  }
+
+  const p = path.normalize(path.join(path.dirname(file), href));
+  if(p.startsWith('../') || p == '..') {
+    messages.push({token, type:'info', message: 'Relative path outside root'});
+    return true;
+  }
+
+  let fullPath = path.join(root, p);
+  if(token.type == 'link') {
+    if(fs.existsSync(fullPath) && fs.statSync(fullPath).isDirectory()) {
+      fullPath = path.join(fullPath, 'index.md');
+    } else if(fullPath.endsWith('.md')) {
+      messages.push({token, type:'warning', message: 'Link should not have a .md extension'});
+    } else {
+      fullPath = fullPath + '.md';
```

Comment:
On the failed CI windows build, is the link checker expecting .png / .jpg links to end in .md?
Handle Tokens.Image?

```
08:22:16 ..../keyman/developer/docs/help/guides/test/keyboard-touch-and-desktop.md - error: Link target 'images/testing/touch-editor-khmer.png.md' does not exist [![Touch Editor -
08:22:16   Khmer](../../images/testing/touch-editor-khmer-800wi.png "Touch Editor - Khmer")](../../images/testing/touch-editor-khmer.png)
08:22:16   ..../keyman/developer/docs/help/guides/test/keyboard-touch-and-desktop.md - error: Link target 'images/testing/touch-editor-build.png.md' does not exist [![Touch Editor - Build
08:22:16   tab](../../images/testing/touch-editor-build-800wi.png "Touch Editor - Build tab")](../../images/testing/touch-editor-build.png)
```

Reply:
Yes, target link exists already, no need to append .md. Fixed.